File: trapezoidal.c
#include <stdio.h>
#include <math.h>

double f(double x) {
    return pow(x, 3);
}

double trap(double a, double b, int n) {
    double dx = (b-a)/n;
    double sum = 0.5 * (f(a) + f(b)); // end points contribution
    
    for (int i=1; i<n; i++) {
        double x = a + i*dx;
        sum += f(x);
    }

    return sum * dx;
}

int main() {
    int n = 1000;
    double a = 0.;
    double b = 1.;

    printf("numerical integration of f(x)=x^3 from a=%.2f to b=%.2f\n",a,b);
    double result = trap(a,b,n);
    double result_an = 0.25*(pow(b,4) - pow(a,4));
    printf("         result=%.4f\n",result);
    printf("analytic result=%.4f\n",result_an);
    printf("      abs error=%.2e\n",fabs(result - result_an));
    return 0;
}

File: trapezoidal.py
def f(x):
    return x**3.

def trap(a,b,n):
    dx = (b-a)/n
    sum = 0.5 * (f(a)+f(b))

    for i in range(1,n):
        x = a + i*dx
        sum += f(x)

    return sum * dx

n = 1000
a = 0.
b = 1.

result = trap(a,b,n)
result_an = 0.25*(b**4-a**4)

print(f"         result={result:.4f}")
print(f"analytic result={result_an:.4f}")
print(f"      abs error={abs(result - result_an):.2e}")